### PR TITLE
inspect memory limit fix

### PR DESCRIPTION
**Pull request by ViugiNick**
Sets a memory limit for execution of inspect (default 10mb)

### CODE REVIEW

File: bin/rdebug-ide [hunk 1]
```diff
@@ -45,6 +51,7 @@ EOB
     options.evaluation_timeout = timeout
   end
   opts.on('--stop', 'stop when the script is loaded') {options.stop = true}
+
```
Comment (valich) [inline bin/rdebug-ide]: Is this change needed here?

File: bin/rdebug-ide [hunk 2]
```diff
@@ -154,4 +161,4 @@ if options.attach_mode
   end
 else
   Debugger.debug_program(options)
-end
```
Comment (valich) [inline bin/rdebug-ide]: formatting

File: lib/ruby-debug-ide/xml_printer.rb [hunk 3]
```diff
@@ -380,9 +384,37 @@ def max_compact_name_size
       50
     end
 
+    def inspect_with_allocation_control(slice, memory_limit)
+      x = Thread.current
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: Can we conceive a better name for this variable? Like `current_thread`, `start_thread` or alike

File: lib/ruby-debug-ide/xml_printer.rb [hunk 4]
```diff
@@ -380,9 +384,37 @@ def max_compact_name_size
...
+      
+      start_alloc_size = ObjectSpace.memsize_of_all
+      
+      trace = TracePoint.new(:c_call, :call) do |tp|
+        curr_alloc_size = ObjectSpace.memsize_of_all
+        
+        if(curr_alloc_size - start_alloc_size > 1e6*memory_limit)
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: spacing

File: lib/ruby-debug-ide/xml_printer.rb [hunk 5]
```diff
@@ -380,9 +384,37 @@ def max_compact_name_size
...
+          
+          trace.disable
+          x.raise MemoryLimitError, "Out of memory: evaluation took longer than 10mb." if x.alive?
+        end
+      end
+
+      trace.enable
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: I think it's better to use `TracePoint#enable( &block)` here

File: lib/ruby-debug-ide/xml_printer.rb [hunk 6]
```diff
@@ -380,9 +384,37 @@ def max_compact_name_size
...
+      slice.inspect
+      trace.disable 
+    rescue MemoryLimitError
+      return nil
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: Do you think it makes sense to add some logging for such cases? Or return something very special

File: lib/ruby-debug-ide/xml_printer.rb [hunk 7]
```diff
@@ -380,9 +384,37 @@ def max_compact_name_size
...
+    end 
+
     def compact_array_str(value)
       slice   = value[0..10]
-      compact = slice.inspect
+      
+      if (defined?(JRUBY_VERSION) || ENV['DEBUGGER_MEMORY_LIMIT'].to_i <= 0)
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: this could be replaced with `compact = if (...)`

File: lib/ruby-debug-ide/xml_printer.rb [hunk 8]
```diff
@@ -425,4 +457,4 @@ def build_value_attr(escaped_value_str)
     
   end
 
-end
+end
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: formatting

File: bin/rdebug-ide (hunk 1 above)
Comment (ViugiNick) [inline bin/rdebug-ide]: fixed

File: bin/rdebug-ide (hunk 2 above)
Comment (ViugiNick) [inline bin/rdebug-ide]: fixed

File: lib/ruby-debug-ide/xml_printer.rb (hunk 3 above)
Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: fixed

File: lib/ruby-debug-ide/xml_printer.rb (hunk 4 above)
Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: fixed

File: lib/ruby-debug-ide/xml_printer.rb (hunk 6 above)
Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: Added logging

File: lib/ruby-debug-ide/xml_printer.rb (hunk 8 above)
Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: fixed

File: lib/ruby-debug-ide/xml_printer.rb (hunk 5 above)
Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: TracePoint#enable( &block) works differently

File: lib/ruby-debug-ide/xml_printer.rb (hunk 7 above)
Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: I do not think that this is applicable in this case

Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: Why so? Please see https://github.com/bbatsov/ruby-style-guide#use-if-case-returns

File: lib/ruby-debug-ide/xml_printer.rb (hunk 5 above)
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: Could you please elaborate more?

Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: `If a block is given, the trace will only be enabled within the scope of the block.`

Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: In my implementation, the checks are carried out more often, which means that the probability of detecting an overflow

Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: Yes, so I suggest changing these three lines with 
```
trace.enable { slice.inspect }
```

> In my implementation, the checks are carried out more often, which means that the probability of detecting an overflow increases

Sorry, did not get what overflow you're talking about.

File: lib/ruby-debug-ide/command.rb [hunk 9]
```diff
@@ -118,7 +118,7 @@ def timeout(sec)
     def debug_eval(str, b = get_binding)
       begin
         str = str.to_s
-        str.force_encoding('UTF-8') if(RUBY_VERSION > 2.0)
+        str.force_encoding('UTF-8') if(RUBY_VERSION > '2.0')
```
Comment (valich) [inline lib/ruby-debug-ide/command.rb]: does this relate to the logging changes?

File: lib/ruby-debug-ide/xml_printer.rb [hunk 10]
```diff
@@ -45,6 +45,7 @@ def initialize(interface)
     end
     
     def print_msg(*args)
+      puts 'def print_msg(*args)'
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: Is this `puts` needed?

File: lib/ruby-debug-ide/xml_printer.rb [hunk 11]
```diff
@@ -413,7 +415,7 @@ def compact_array_str(value)
       compact = if (defined?(JRUBY_VERSION) || ENV['DEBUGGER_MEMORY_LIMIT'].to_i <= 0)
                   slice.inspect
                 else  
-                  compact = inspect_with_allocation_control(slice, ENV['DEBUGGER_MEMORY_LIMIT'].to_i)
+                  compact = inspect_with_allocation_control(slice, ENV['DEBUGGER_MEMORY_LIMIT'].to_i, value.object_id.to_s)
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: I am not sure `object_id` us really useful for logging purposes: it does not tell anything about the object. I think the previous message already a good start.

File: lib/ruby-debug-ide/xml_printer.rb [hunk 12]
```diff
@@ -404,6 +405,7 @@ def inspect_with_allocation_control(slice, memory_limit)
       trace.disable
       result 
     rescue MemoryLimitError => e
+      print_msg(e.message)
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: I think one may make use of `--debug` option here:
```
opts.on("-d", "--debug", "Debug self - prints information for debugging ruby-debug itself") do
     Debugger.cli_debug = true
     options.cli_debug = true
   end
```

File: lib/ruby-debug-ide/xml_printer.rb (hunk 6 above)
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: I think printing also a stacktrace of an exception may help investigating the problems in `inspect` calls for users.

Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: This backtrack will end with the calls of our debase and r-d-i methods. I think that this is not very useful.

File: bin/rdebug-ide [hunk 13]
```diff
@@ -37,8 +37,8 @@ EOB
   opts.separator ""
   opts.separator "Options:"
   
-  ENV['DEBUGGER_MEMORY_LIMIT'] = '10'
-  opts.on("-m", "--memory-limit LIMIT", Integer, "evaluation memory limit in mb (default: 10)") do |limit| 
+  ENV['DEBUGGER_MEMORY_LIMIT'] = '1'
```
Comment (valich) [inline bin/rdebug-ide]: What stands behind such defaults change? Could you please reason it?

File: lib/ruby-debug-ide/xml_printer.rb [hunk 14]
```diff
@@ -385,28 +385,33 @@ def max_compact_name_size
...
-      trace = TracePoint.new(:c_call, :call) do |tp|
-        curr_alloc_size = ObjectSpace.memsize_of_all
-        
-        if(curr_alloc_size - start_alloc_size > 1e6 * memory_limit)
-          
-          trace.disable
-          curr_thread.raise MemoryLimitError, "Out of memory: evaluation of inspect took more than #{memory_limit}mb." if curr_thread.alive?
-        end
+      if Thread.respond_to?(:critical) and Thread.critical
+        raise ThreadError, "memory_limit within critical session"
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: AFAIU the execution of that line does not mean memory limit is exceeded. What does this check?

File: lib/ruby-debug-ide/xml_printer.rb [hunk 15]
```diff
@@ -385,28 +385,33 @@ def max_compact_name_size
...
-      trace.enable
-      result = slice.inspect
-      trace.disable
-      result 
+      result = nil
+      inspect_thread = DebugThread.start {
+        start_alloc_size = ObjectSpace.memsize_of_all
+        trace = TracePoint.new(:c_call, :call) do |tp|
+          curr_alloc_size = ObjectSpace.memsize_of_all
+          start_alloc_size = curr_alloc_size if (curr_alloc_size < start_alloc_size)
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: The case of GC?

File: lib/ruby-debug-ide/xml_printer.rb [hunk 16]
```diff
@@ -385,28 +385,33 @@ def max_compact_name_size
...
+      curr_thread = Thread.current
 
-      trace.enable
-      result = slice.inspect
-      trace.disable
-      result 
+      result = nil
+      inspect_thread = DebugThread.start {
+        start_alloc_size = ObjectSpace.memsize_of_all
+        trace = TracePoint.new(:c_call, :call) do |tp|
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: Do you need `trace` variable? I don't see any usages..

File: lib/ruby-debug-ide/xml_printer.rb [hunk 17]
```diff
@@ -385,28 +385,33 @@ def max_compact_name_size
...
+          curr_alloc_size = ObjectSpace.memsize_of_all
+          start_alloc_size = curr_alloc_size if (curr_alloc_size < start_alloc_size)
+
+          if(curr_alloc_size - start_alloc_size > 1e6 * memory_limit)
+            curr_thread.raise MemoryLimitError, "Out of memory: evaluation of inspect took more than #{memory_limit}mb. \n#{caller.map{|l| "\t#{l}"}.join("\n")}"
+            inspect_thread.kill
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: I'd rather move this kill to line 410, after join. It seems strange to me to kill the thread from this thread itself.

File: lib/ruby-debug-ide/xml_printer.rb (hunk 15 above)
Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: exactly

File: lib/ruby-debug-ide/xml_printer.rb (hunk 16 above)
Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: I'll disable it throw 'trace'

File: lib/ruby-debug-ide/xml_printer.rb (hunk 17 above)
Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: fixed

File: bin/rdebug-ide (hunk 13 above)
Comment (ViugiNick) [inline bin/rdebug-ide]: With 10 you have to wait for a pretty long time

File: lib/ruby-debug-ide/xml_printer.rb [hunk 18]
```diff
@@ -389,9 +392,19 @@ def inspect_with_allocation_control(slice, memory_limit)
       result = nil
       inspect_thread = DebugThread.start {
         start_alloc_size = ObjectSpace.memsize_of_all
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: Overall looks nice. One last question: do you think such checks on ___every___ call will slow down the execution?
The possible solution there is to make these checks only in 10% of calls (like `Math.random > 0.9`) or alike

Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: Most often 'inspect' works fast and requires little memory, so in general the performance should not slow down. And in case of heavy inspect-s, when using probability, everything can crush with a high probability. Then why was this fix at all?)

Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: Perhaps, it is worth running a :call and :c_call handler a little less often. Every fifth, for example.

Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: > Perhaps, it is worth running a :call and :c_call handler a little less often. Every fifth, for example.

That's what I'm talking about.

`inspect` often works fast, but
1. We add constant overhead on each call which may slow execution down in dozens of times;
2. There are many `inspects` during variables collection so in total this matters

Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: I sill getting an INTERNAL ERROR (pipe or something)  when using probabilistic scheme. Apparently, it is better to abandon this proposal.

Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: Can you show the snippet you're using to achieve the result?

Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: TracePoint is never lucky(

Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: `trace = TracePoint.new(:c_call, :call) do |tp|
         
          if(Math.random > 0.75)
            curr_alloc_size = ObjectSpace.memsize_of_all
            curr_time = Time.now
           
            if(curr_time - start_time > max_time)
              curr_thread.raise TimeLimitError, "Timeout: evaluation of inspect took longer than #{max_time}sec. \n#{caller.map{|l| "\t#{l}"}.join("\n")}"
              trace.disable
            end
 
            start_alloc_size = curr_alloc_size if (curr_alloc_size < start_alloc_size)
           
            if(curr_alloc_size - start_alloc_size > 1e6 * memory_limit)
              curr_thread.raise MemoryLimitError, "Out of memory: evaluation of inspect took more than #{memory_limit}mb. \n#{caller.map{|l| "\t#{l}"}.join("\n")}"
              trace.disable
            end
          end
        end.enable {
          result = slice.inspect
        }`

Comment (ViugiNick) [inline lib/ruby-debug-ide/xml_printer.rb]: fixed

File: lib/ruby-debug-ide/xml_printer.rb [hunk 19]
```diff
@@ -380,13 +387,57 @@ def max_compact_name_size
       50
     end
 
+    def inspect_with_allocation_control(slice, memory_limit)
+      curr_thread = Thread.current
+      result = nil
+      inspect_thread = DebugThread.start {
+        start_alloc_size = ObjectSpace.memsize_of_all
+        start_time = Time.now
+        max_time = Debugger.evaluation_timeout
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: I think 10s evaluation timeout is too much for this place. The reason is that evaluation is an explicit user action so it's OK to try to wait for a possibly long evaluation. In the case of convenience infos calculated dozens of times per breakpoint, 10s does not make much sense to me. I'd say we should make it somewhere between 100 and 500ms.

File: lib/ruby-debug-ide/xml_printer.rb [hunk 20]
```diff
@@ -144,7 +158,46 @@ def print_string(string)
         print_variable('encoding', string.encoding, 'instance') if string.respond_to?('encoding')         
       end
     end
-    
+
+    def exec_with_allocation_control(value, memory_limit, time_limit, exec_method, flag)
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: what does `flag` mean? Not a helpful name for a parameter

File: lib/ruby-debug-ide/xml_printer.rb [hunk 21]
```diff
@@ -6,10 +6,24 @@
 module Debugger
 
   class MemoryLimitError < StandardError  
+    attr_reader :message
+    attr_reader :backtrace
+
+    def initialize(message, backtrace)
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: This one does not provide default `backtrace` value and the one below does. Is that intentional?

File: lib/ruby-debug-ide/xml_printer.rb [hunk 22]
```diff
@@ -144,7 +158,46 @@ def print_string(string)
...
+          result = value.send exec_method
+        }
+      }
+      inspect_thread.join
+      inspect_thread.kill
+      return result
+    rescue MemoryLimitError, TimeLimitError => e
+      print_debug(e.message + "\n" + e.backtrace)
+      
+      return nil if flag
```
Comment (valich) [inline lib/ruby-debug-ide/xml_printer.rb]: ternary

**Comment by valich** [comment]
Squashed and merged into master.
